app/services/onlinetesting/Phase3TestService.scala

```diff
@@ -217,20 +217,31 @@ trait Phase3TestService extends OnlineTestService with Phase3TestConcern {
       }
     }
 
-  def markAsCompleted(launchpadInviteId: String)(implicit hc: HeaderCarrier, rh: RequestHeader): Future[Unit] = eventSink {
-    phase3TestRepo.getTestGroupByToken(launchpadInviteId).flatMap { test =>
-      if (test.testGroup.tests.find(_.token == launchpadInviteId).get.completedDateTime.isEmpty) {
-        for {
-          _ <- phase3TestRepo.updateTestCompletionTime(launchpadInviteId, dateTimeFactory.nowLocalTimeZone)
-          updated <- phase3TestRepo.getTestGroupByToken(launchpadInviteId)
-          _ <- phase3TestRepo.updateProgressStatus(updated.applicationId, ProgressStatuses.PHASE3_TESTS_COMPLETED)
-        } yield {
-          AuditEvents.VideoInterviewCompleted(updated.applicationId) ::
-            DataStoreEvents.VideoInterviewCompleted(updated.applicationId) ::
-            Nil
+  def markAsCompleted(launchpadInviteId: String)(implicit hc: HeaderCarrier, rh: RequestHeader): Future[Unit] = {
+
+    // Launchpad only: If a user has completed unexpire them
+    def removeExpiryIfSet(applicationId: String, expired: Boolean) = expired match {
```

Comment:
I'm not so keen on this removeExpiryIfSet method if I'm being picky. This is a form of control coupling (passing the boolean flag) that isn't necessary.  If you rename the method to removeExpiry and make the decision in the parent code whether to call it or not you will lessen the coupling.